### DISCUSSION TITLE
Visual Studio output format

### DISCUSSION THREAD

**Discussion by xrouge** [BurntSushi/ripgrep#1994] [General] [answered]
Uniquely to VS output is the formatting of the line number between (). This allows double-clicking on the output line in order to open the file and go to the  line number. For example:

`myfile.cpp(32) : Hello World`

It was suggested to pipe the regular output to 'sed', or format as json and parse outpur, but it may be easier to implement this internally. There is an option to replace the ':', but it's not sufficient because VS needs a different opening and closing character.

**Comment by BurntSushi** [comment, marked as answer]
See:

* https://github.com/BurntSushi/ripgrep/issues/665
* https://github.com/BurntSushi/ripgrep/issues/1326
* https://github.com/BurntSushi/ripgrep/issues/700
* https://github.com/BurntSushi/ripgrep/issues/390

TL;DR - #665 is a possible way of solving this, but I don't think it's ripgrep's place to provide infinite flexibility in its output format. Tools like Visual Studio should consider adapting to the grep format convention that has existed for decades.

**Reply by ssbarnea** [reply]
@BurntSushi I do appreciate your work a lot, but the reality is that #665 is misleading and not fix or even workaround.

How about finding some common ground? I remember moving from ag to rg due to being unpleased with project maintenance few years back and I would hate to do it again.

That is not even about my personal need, as in the end I could write a wrapper to sanitize the output of ripgrep and workaround it. That is more about welcoming contributions from others and making the project more welcoming to external contributions. If people are refused contributions like this that fix real world-usage, they will be unlikely to contribute later bug-fixes or just help others use the tool.

As I was involved with several vscode issues related to terminal, I can tell you that things will not change soon there. Their terminal does not support hyperlink extensions and that is unlikely to change any time soon, basically due to a combination of external dependency and security concerns. Some people forget that vscode also has a browser version, the one that starts when you press `.` key on a github repo page. That has implications when it comes to links.

So instead of throwing the dead cat to other projects, how about finding a common ground and allow users to customize the output? Having a format line should be easy to add and likely not need further maintenance. That is even more important if we count that the current output format is not pep8 compatible: `filename:line:column: message` and that format is almost industry standard for any linter.

I can understand that you do not want to change the default output, for good reasons. Still, enabling others to tune would not affect/break current users.

Considering that you would not have to implement it yourself and you can have high standards regarding how to be implemented, is there any chance to revisit your decision?

## hour later...

I found myself replacing the existing rg alias used to overcome other limitations with a considerably more complex shell function, as aliases do not work well with pipeling, only to change the output to make it pep8 compatible.

```shell
rg () {
  command rg --hidden --no-follow --max-columns 255 --no-heading --column -F --json "$@" | \
  jq -j -r 'select(.type|contains("match"))| .data.path.text + ":" + (.data.line_number | tostring) + " " + .data.lines.text';
}
```

I mentioned this here, as others might find it useful workaround, hopefully not a permanent one.

**Reply by BurntSushi** [reply]
In your work-around, `--max-columns`, `--no-heading` and `--column` are having no impact. You can remove all of them. You can probably also remove `--no-follow` since that's the default, unless you have `-L/--follow` set in a config file somewhere.

As to the rest of your comment, I don't really appreciate it and I don't find your approach here particularly friendly. I resent the implication that saying "no" to a feature request is equivalent to being less welcoming. And other aspects of your comment aren't particularly nice either. For example, I find this to be passive aggressive:

> I remember moving from ag to rg due to being unpleased with project maintenance few years back and I would hate to do it again.

Anyway, moving on...

> I can understand that you do not want to change the default output, for good reasons. Still, enabling others to tune would not affect/break current users.

ripgrep _does_ permit tuning your output. It just isn't flexible enough to do what folks want to work around limitations in other software. Presenting the status quo as "no changes to the default output" makes it appear a lot more unreasonable than it actually is.

> That is even more important if we count that the current output format is not pep8 compatible: `filename:line:column: message` and that format is almost industry standard for any linter.

I don't know what this means. Doesn't `pep8` refer to a Python style guide? I don't know what its relevance to this issue is.

> I do appreciate your work a lot, but the reality is that #665 is misleading and not fix or even workaround.

Why is it misleading? Why won't it fix things? Don't leave me hanging! I think you later say that VS doesn't support links in the terminal? That seems like important information to add to that issue that nobody has added yet. I still don't understand why you're saying it's misleading though. And given that nobody has said anything about this at all and given that #665 is still open, I find it quite curious that I'm being accused of "throwing a dead cat" to other projects. It might not have occurred to you that I didn't know that links wouldn't work in Visual Studio.

In any case, the best way to advance something like this is to provide a concrete proposal in a new issue that solves the problem you care about. I personally care a lot of maintenance and the surface area of ripgrep. The surface area of ripgrep is already _very large_, and I want to be very careful in how it gets expanded.

In case it isn't clear: I have literally never used Visual Studio in my entire life. You and others may have considerable context that I don't have. Folks will have to work harder to make that context clearer. Similarly, I haven't used Windows in any serious capacity in well over a decade at this point, so if there are Windows specific things in play here, those need to be made explicit. (I do have a Windows laptop that I'll begrudgingly drag out for testing, but that's it.)

**Reply by ssbarnea** [reply]
"visual studio" and [vscode](https://code.visualstudio.com/) (also known as `code`) are two totally different things. The first one is the Windows one and I cannot speak about it because I did not touch it in over a decade. On the other hand vscode is based on electron/atom, multiplatform and (mostly) open source. In fact there is a pure open-source version of it named https://vscodium.com/ and from what they do/support, you can find find the almost the same.

Yep, `pep8` is both a Python standard but used to be a tool which later was renamed to pycodestyle. I referred to it because in python world people alre refer to pep8 tool output format as being "standard", basically because most tool authors wanted to produce output that can be recognized by terminals or IDEs, allowing people to click those errors lines to open them into their editor.

So basically I used `pep8` as a shortcut for describing `filename:line:column: message` and the simplified `filename:line: message` version which has no column part, as column is not always available.

I can even present a list of tools that are able to produce these by default or with optional formatting parameter: pycodestyle, flake8, yamllint, doc8, pylint. 

Re my example, I removed the extra lines, they were leftovers from the time I was not using `jq` to format the output.

Back to practical things: the only thing that we need is to add one space before the line content to make it work. It is bit hilarious because click-to-open works on line that have spacing at their start.

@xrouge Can you please report if "visual studio" recognizes the `filename:line: message` and `filename:line:column message` messages? If so, we maybe conclude that this format is the most popular/compatible one. I can try do do it myself, once I got onto a Windows box.

**Reply by BurntSushi** [reply]
> "visual studio" and [vscode](https://code.visualstudio.com/) (also known as `code`) are two totally different things. The first one is the Windows one and I cannot speak about it because I did not touch it in over a decade.

But that's what this _particular_ issue is about! Unless the OP meant to say "VS Code" instead of "VS," this issue is about Visual Studio. So yeah, if _you're_ talking about VS Code, then that's almost certainly completely unrelated to this specific discussion post. So now I'm confused as to why we're having this conversation _here_...

I would encourage you to contribute to #665. Because it sounds to me like you're saying VS Code doesn't support hyperlinks and thus that issue doesn't solve the problem it intends to solve. So we need to get to the bottom of that.

**Comment by jplimack** [comment]
```bash
if [[ "$TERM_PROGRAM" == 'vscode' ]]; then
  alias 'rg'='rg --smart-case --hidden --no-heading --column'
else
  alias 'rg'='rg --smart-case --hidden'
fi
```

command+click'ing on `internal/server/server.go:138:15` takes me straight there in vscode. 
```bash
rg Registered
internal/server/server.go:138:15:       log.Println("Registered routes:")
```

From https://www.hardscrabble.net/2021/ripgrep-in-vscode/
